code/modules/power/pt_laser.dm

```diff
@@ -339,158 +337,96 @@
-
-	t += "<BR><br>lifetime earnings:<br>[engineering_notation(lifetime_earnings)] credits</PRE><HR><A href='?src=\ref[src];close=1'>Close</A>"
-
-	t += "</TT>"
-	user.Browse(t, "window=Power Transmission Laser;size=460x300")
-	onclose(user, "Power Transmission Laser")
-	return
-
-/obj/machinery/power/pt_laser/Topic(href, href_list)
-	..()
-
-	if (usr.stat || usr.restrained() )
+/obj/machinery/power/pt_laser/ui_state(mob/user)
+	return tgui_default_state
+
+/obj/machinery/power/pt_laser/ui_status(mob/user, datum/ui_state/state)
+	return min(
+		state.can_use_topic(src, user),
+		tgui_broken_state.can_use_topic(src, user),
+		tgui_not_incapacitated_state.can_use_topic(src, user)
+	)
+
+/obj/machinery/power/pt_laser/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PowerTransmissionLaser")
+		ui.open()
+
+/obj/machinery/power/pt_laser/ui_data(mob/user)
+	var/list/data = list()
+	data["capacity"] = src.capacity
+	data["charge"] = src.charge
+	data["chargingEnabled"] = src.charging
+	data["excessPower"] = src.excess
+	data["gridLoad"] = src.terminal.powernet.load
+	data["inputLevel"] = src.chargelevel
+	data["inputMultiplier"] = src.input_multi
+	data["inputNumber"] = src.input_number
+	data["isCharging"] = src.is_charging
+	data["isFiring"] = src.firing
+	data["laserEnabled"] = src.online
+	data["lifetimeEarnings"] = src.lifetime_earnings
+	data["name"] = src.name
+	data["outputLevel"] = src.output
+	data["outputMultiplier"] = src.output_multi
+	data["outputNumber"] = src.output_number
+	data["totalGridPower"] = src.terminal.powernet.avail
+	return data
+
+/obj/machinery/power/pt_laser/ui_act(action, params)
+	if(..())
 		return
-
-	if (( usr.using_dialog_of(src) && ((get_dist(src, usr) <= 1) && istype(src.loc, /turf))) || (isAI(usr)))
-		if( href_list["close"] )
-			usr.Browse(null, "window=Power Transmission Laser")
-			src.remove_dialog(usr)
-			return
-
-		else if( href_list["cmode"] )
-			charging = !charging
-			updateicon()
-
-		else if( href_list["online"] )
-			online = !online
+	switch(action)
+		//Input controls
+		if("toggleInput")
+			src.charging = !src.charging
+			. = TRUE
+		if("setInput")
+			src.input_number = clamp(params["setInput"], 0, 999)
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
+		if("inputW")
+			src.input_multi = 1 WATT
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
+		if("inputkW")
+			src.input_multi = 1 KILO WATT
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
+		if("inputMW")
+			src.input_multi = 1 MEGA WATT
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
+		if("inputGW")
+			src.input_multi = 1 GIGA WATT
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
+		if("inputTW")
+			src.input_multi = 1 TERA WATT
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
```

Comment:
Future reference, could do as a single action with params, but I get that you'd then have to convert the param into a number anyway so NBD here. It's neat enough.